include/posix/pthread.h

```diff
@@ -383,4 +385,9 @@ int pthread_barrierattr_getpshared(const pthread_barrierattr_t *, int *);
 int pthread_barrierattr_setpshared(pthread_barrierattr_t *, int);
 */
 
+/* Sleep Related APIs */
+unsigned int sleep(unsigned int seconds);
+unsigned int msleep(unsigned int mseconds);
+int usleep(useconds_t useconds);
```

Comment:
According to POSIX (http://pubs.opengroup.org/onlinepubs/7908799/xsh/usleep.html), usleep is declared in unistd.h, why these are being put here? This can lead to header conflicts.

Reply:
Since Zephyr supports more than 1 C library, I believe that the best way to solve this problem is to use `#include_next`.

This isn't a standard C directive, but it's supported by GCC, Clang, and other major compilers.  The build system is then modified so that, if POSIX compatibility is required a `-I /path/to/posix/compat/headers` is passed to the compiler before any other include directory path.  Each header can then have something like:

```C
#include_next <unistd.h>

#ifndef UNISTD_POSIX_COMPAT
#define UNISTD_POSIX_COMPAT

#if defined(CONFIG_POSIX_COMPAT)

int usleep(useconds_t useconds);

#endif

#endif /* UNISTD_POSIX_COMPAT */
```

This way, programs using `#include <unistd.h>` will include this file first, which in turn will include the C library's `unistd.h` -- and then the POSIX compatibility stuff will be declared.

Reply:
> This way, programs using #include <unistd.h> will include this file first, which in turn will include the C library's unistd.h

... Or maybe we should rely on C library's unistd.h in the first place (if it exists). As I mention above, POSIX compliance is more than just code which implements POSIX functions. In the end of the line, POSIX compliance is that any POSIX compliant app should be buildable without any adhoc `-I` switches - it's just  if an app uses POSIX headers, it should just work (and non-POSIX headers is what should be namespaced away).

All in all, we can either think about things like that beforehand, and achieve more or less no-nonsense POSIX compliance in 4-5 iterations, or do it in "random" manner, and go thru dozens of big refactors until we get it right.